apollo-federation/src/sources/connect/json_selection/methods.rs

```diff
@@ -0,0 +1,2178 @@
+use apollo_compiler::collections::IndexMap;
+use apollo_compiler::collections::IndexSet;
+use lazy_static::lazy_static;
+use serde_json_bytes::ByteString;
+use serde_json_bytes::Map as JSONMap;
+use serde_json_bytes::Value as JSON;
+
+use super::helpers::json_type_name;
+use super::immutable::InputPath;
+use super::lit_expr::LitExpr;
+use super::ApplyTo;
+use super::ApplyToError;
+use super::MethodArgs;
+use super::PathList;
+use super::VarsWithPathsMap;
+
+type ArrowMethod = fn(
+    // Method name
+    method_name: &str,
+    // Arguments passed to this method
+    method_args: &Option<MethodArgs>,
+    // The JSON input value (data)
+    data: &JSON,
+    // The variables
+    vars: &VarsWithPathsMap,
+    // The input_path (may contain integers)
+    input_path: &InputPath<JSON>,
+    // The rest of the PathList
+    tail: &PathList,
+    // Errors
+    errors: &mut IndexSet<ApplyToError>,
+) -> Option<JSON>;
+
+lazy_static! {
+    // This set controls which ->methods are exposed for use in connector
+    // schemas. Non-public methods are still implemented and tested, but will
+    // not be returned from lookup_arrow_method outside of tests.
+    static ref PUBLIC_ARROW_METHODS: IndexSet<&'static str> = {
```

Comment:
@benjamn did you look into using an `enum` for these instead? I just wonder if it might make validations easier when we want to check for specific methods. That would also mean it won't be possible for us to mismatch a definition here with its implementation down below, since there would be a `ArrowMethod::execute` or similar which we have to fill in on a `match`.

If the `IndexSet` & `IndexMap` are more ergonomic, we should probably switch to constants for their names so we can reuse them elsewhere as well.

Reply:
But we'd still have to map the method name string (from the AST) to one of those enum values, just to look it up in this set. That seems like an unnecessary layer of indirection.

If we're using enums for all method names (public and private), you could still accidentally include or forget to include one of the enums in the set of public methods.

I think ultimately we want the registered method definition to be an object implementing a trait with methods like `apply` and `is_public()` (not just a function). Then `lookup_arrow_method` wouldn't have to consult two maps, but could instead check `is_public()` for the retrieved object before returning it.

Reply:
I'm going to try my hand at whatever refactor works best when I actually write the validations & can test out the ergonomics. So nothing more to do on this in this PR.